### PR TITLE
Use compiler builtins

### PR DESCRIPTION
**Pull request by lvella**
Instead of rolling out our own. Address issue #268 .

### CODE REVIEW

File: riscv/src/reachability.rs [hunk 1]
```diff
@@ -89,11 +92,42 @@ fn extract_replacements(statements: &[Statement]) -> BTreeMap<&str, &str> {
             _ => None,
         })
         .fold(BTreeMap::new(), |mut acc, (from, to)| {
-            if acc.insert(from, to).is_some() {
+            if acc.insert(from, Cell::new(to)).is_some() {
```
Comment (chriseth) [inline riscv/src/reachability.rs:95]: Do we really need Cell for that?

Comment (lvella) [reply riscv/src/reachability.rs:95]: Well, there are many ways to do it. But to mutate the tree in-place in a depth search to ensure that each transition is only walked once, this is the least ugly way to do it.

Comment (lvella) [reply riscv/src/reachability.rs:95]: To make this solution perfect, we just need to write a rust RFC to change `BTreeMap` interface to provide a way to convert a `BTreeMap<&str, Cell<&str>>` to a `BTreeMap<&str, &str>` without rebuilding the entire tree, as both are binary-identical.

Comment (chriseth) [reply riscv/src/reachability.rs:95]: Why do you need to modify it in place? Wouldn't it be easier to just create a copy?

Comment (lvella) [reply riscv/src/reachability.rs:95]: No! Wait!

Maybe there is a way to do it if I create a Vec with the keys before doing the depth search.

Comment (lvella) [reply riscv/src/reachability.rs:95]: > Why do you need to modify it in place? Wouldn't it be easier to just create a copy?

Yes, but consider the case where a long path is walked to find the final name of a symbol. For every intermediate name in the path, a fraction of this same path will be walked to find the final name, which makes the algorithm O(N² log N). If I change the final name in-place as I walk the path the first time, the algorithm is then O(N log N).

Of course, this is completely irrelevant unless you expect someone to try a DoS attack on the compiler.

Comment (lvella) [reply riscv/src/reachability.rs]: Changed to the idea that collect the keys into a Vec. It is much more beautiful now. Still inplace and O(N log N).

**Comment by chriseth** [comment]
Nice! Can you squash this please?

**Comment by lvella** [comment]
> Nice! Can you squash this please?

Like this? Why not just use the github button "Squash and merge" ?

**Pull request merged**